### PR TITLE
Issue #20412: Adding test to validate SAML with distributed auth cache.

### LINKED ISSUE #20412: SEC 380: Test SAML with distributed authentication cache
Perform distributed authentication cache testing with the samlWeb-2.0 feature.

### PR DESCRIPTION
**Pull request by jvanhill**
fixes #20412

### CODE REVIEW

**Comment by jvanhill** [comment]
#build

**Comment by jvanhill** [comment]
#build

**Comment by jvanhill** [comment]
#build

**Comment by tevans78** [comment]
> **The following files contain a string indicating discriminating content:**
> 
>     * dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/docker/KeycloakContainer.java
> 
>     * dev/io.openliberty.jcache.internal_fat.hazelcast/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.1/server.xml
> 
>     * dev/io.openliberty.jcache.internal_fat.hazelcast/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.2/server.xml

There is no way around using the word "master" in this context as it is the default realm name which is baked into the Keycloak docker image. https://www.keycloak.org/getting-started/getting-started-docker#_create_a_realm

**Comment by jvanhill** [comment]
#build
#libby

**Comment by jvanhill** [comment]
See the comment earlier in this PR by @tevans78 regarding these.

> **The following files contain a string indicating discriminating content:**
> 
> * dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.2/server.xml
> * dev/io.openliberty.jcache.internal_fat.hazelcast/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.2/server.xml
> * dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/docker/KeycloakContainer.java
> * dev/io.openliberty.jcache.internal_fat.hazelcast/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.1/server.xml
> * dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.1/server.xml

**Comment by jvanhill** [comment]
#build
#libby

**Comment by jvanhill** [comment]
See the comment earlier in this PR by @tevans78 regarding these.

> **The following files contain a string indicating discriminating content:**
> 
> * dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.2/server.xml
> * dev/io.openliberty.jcache.internal_fat.hazelcast/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.2/server.xml
> * dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/docker/KeycloakContainer.java
> * dev/io.openliberty.jcache.internal_fat.hazelcast/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.1/server.xml
> * dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.1/server.xml

**Comment by jvanhill** [comment]
There appears there was a network outage in the build system. Re-building.

#build

**Comment by kristip17** [comment]
Review: Discussed interop/versioning on read/write objects (follow up with Jared), externalizing strings being thrown (maybe there's a generic message that could be reused for the unlikely-to-occur exceptions), commenting the use of master in the test (follow up with ops team), and adding more info trace on the new FAT.

**Review by kristip17 — APPROVED** [review]

**Comment by jvanhill** [comment]
#build
#libby

**Comment by jvanhill** [comment]
See the comment earlier in this PR by @tevans78 regarding these.

> **The following files contain a string indicating discriminating content:**
> 
> * dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.2/server.xml
> * dev/io.openliberty.jcache.internal_fat.hazelcast/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.2/server.xml
> * dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/docker/KeycloakContainer.java
> * dev/io.openliberty.jcache.internal_fat.hazelcast/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.1/server.xml
> * dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.1/server.xml

**Comment by jvanhill** [comment]
FAT failures from latest build were existing / known failures.

**Review by kristip17 — APPROVED** [review]

**Review by arkarkala — APPROVED** [review]

**Comment by arunavemulapalli** [comment]
my review is limited to saml sso / websso changes that we've looked at yesterday.
I will be removing the saml sso project/bundle soon (once we complete the jaxws-2.2 feature update)
saml websso (which is using opensaml v3) should be the default for samlWeb feature.

**Review by arunavemulapalli — APPROVED** [review]

File: dev/com.ibm.ws.security.authentication/src/com/ibm/ws/security/authentication/cache/CacheObject.java [hunk 1]
```diff
@@ -59,7 +69,62 @@ public void addLookupKey(Object key) {
...
+     * principals and credentials of the {@link Subject} since the SecureSet implementation was not intended for serialization.
+     *
+     * @param <T> The type for the set to copy.
+     * @param set The set to copy.
+     * @return The copied set.
+     */
+    private static <T> Set<T> copySet(Set<T> set) {
+        Set<T> copy = new HashSet<T>();
+        copy.addAll(set);
+        return copy;
```
Comment (jhanders34) [inline dev/com.ibm.ws.security.authentication/src/com/ibm/ws/security/authentication/cache/CacheObject.java]:
```suggestion
        return new HashSet<T>(set);
```
By calling this constructor it does the sizing better.

File: dev/com.ibm.ws.security.authentication/src/com/ibm/ws/security/authentication/cache/CacheObject.java [hunk 2]
```diff
@@ -59,7 +69,62 @@ public void addLookupKey(Object key) {
...
+    private void writeObject(ObjectOutputStream output) throws IOException {
+        /*
+         * Write all non-transient fields.
+         */
+        output.defaultWriteObject();
+
+        /*
+         * Write the fields comprising the Subject.
+         */
+        output.writeBoolean(subject.isReadOnly());
```
Comment (jhanders34) [inline dev/com.ibm.ws.security.authentication/src/com/ibm/ws/security/authentication/cache/CacheObject.java:126]: Should this serialization have a versioning option?

File: dev/com.ibm.ws.security.saml.sso_fat.common/test-applications/samlclient.war/resources/META-INF/permissions.xml [hunk 3]
```diff
@@ -45,5 +45,11 @@
       <actions>read</actions>
    </permission>
    
+   <permission>
```
Comment (jhanders34) [inline dev/com.ibm.ws.security.saml.sso_fat.common/test-applications/samlclient.war/resources/META-INF/permissions.xml:48]: Why is this needed?  Are we missing a doPriv with how we are doing the serialization?

Comment (jvanhill) [reply dev/com.ibm.ws.security.saml.sso_fat.common/test-applications/samlclient.war/resources/META-INF/permissions.xml:48]: I tried adding a doPriv where it failed in the application, but it did not resolve the issue so instead I added here like others have done. I am not quite sure why the doPriv didn't resolve it.

File: dev/com.ibm.ws.security.authentication/src/com/ibm/ws/security/authentication/cache/CacheObject.java (hunk 2 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.authentication/src/com/ibm/ws/security/authentication/cache/CacheObject.java:126]: I added Serialization in this class for the Beta. I found during testing for this change that I had messed it up. I could version this, but I would only be versioning it to work with a beta release, and instead I intend on future beta releases to recommend clearing the cache before running a new version.

File: dev/com.ibm.ws.security.authentication/src/com/ibm/ws/security/authentication/cache/CacheObject.java (hunk 1 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.authentication/src/com/ibm/ws/security/authentication/cache/CacheObject.java]: Removed the copySet method and just called the constructor instead.

File: dev/com.ibm.ws.security.authentication/src/com/ibm/ws/security/authentication/cache/CacheObject.java (hunk 2 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.authentication/src/com/ibm/ws/security/authentication/cache/CacheObject.java:126]: I added a serialization version field suitable for future use in the readObject method should serialization change down the road.

**Review by jhanders34 — APPROVED** [review]

File: dev/com.ibm.ws.security.saml.sso_fat.common/test-applications/samlclient.war/resources/META-INF/permissions.xml (hunk 3 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.saml.sso_fat.common/test-applications/samlclient.war/resources/META-INF/permissions.xml:48]: Opened an issue to investigate this in the future: https://github.com/OpenLiberty/open-liberty/issues/20560

**Comment by jvanhill** [comment]
#build
#libby

**Comment by jvanhill** [comment]
See the comment earlier in this PR by @tevans78 regarding these.



> **The following files contain a string indicating discriminating content:**
> 
> * dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.2/server.xml
> * dev/io.openliberty.jcache.internal_fat.hazelcast/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.2/server.xml
> * dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/docker/KeycloakContainer.java
> * dev/io.openliberty.jcache.internal_fat.hazelcast/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.1/server.xml
> * dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.1/server.xml

**Comment by jvanhill** [comment]
#build

**Comment by jvanhill** [comment]
Respun the build. There were failures that I don't think are attributable to this change, but running again just to verify.

https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_oTtDAKt9EeyJdcarUSDuUg

**Comment by jvanhill** [comment]
#build
#libby

**Comment by jvanhill** [comment]
#build

**Comment by jvanhill** [comment]
There is some build instability, so I am requesting a rebuild yet again.

#build

**Comment by jvanhill** [comment]
#libby

**Comment by jvanhill** [comment]
#libby

**Comment by jvanhill** [comment]
See the comment earlier in this PR by @tevans78 regarding these.

> **The following files contain a string indicating discriminating content:**
> 
> * dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.2/server.xml
> * dev/io.openliberty.jcache.internal_fat.hazelcast/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.2/server.xml
> * dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/docker/KeycloakContainer.java
> * dev/io.openliberty.jcache.internal_fat.hazelcast/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.1/server.xml
> * dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.saml.auth.cache.1/server.xml
